torch/nested/_internal/nested_tensor.py

```diff
@@ -83,6 +112,28 @@ def __repr__(self):
             grad_fn_str = f", grad_fn={self.grad_fn}"
         return f"NestedTensor(size={self._size}, offsets={self.offsets}{grad_fn_str})"
 
+    def __tensor_flatten__(self):
+        return ["_values", "_offsets"], (self.requires_grad, self._size[self._ragged_idx])
+
+    @staticmethod
+    def __tensor_unflatten__(inner_tensors: Dict, meta):
+        assert len(inner_tensors) == 2
+        values = inner_tensors["_values"]
+        offsets = inner_tensors["_offsets"]
+        (requires_grad, ragged_size, *extra_meta) = meta
+
+        if len(extra_meta) > 0:
```

Comment:
The extra_meta isn't actually optional here, right? Since you always need one to properly reconstruct the subclass

I think I would expect that `__tensor_flatten__()` returns the ragged dimension as another piece of metadata, that way we can plumb it back to this `__tensor_unflatten__()` call when we reconstruct the JaggedTensor.

And then in the fakeifying case, the ragged dimension SymInt is actually the "wrong" SymInt, since it's the one from eager mode, and not the dynamo one we need that actually has a source attached.

Reply:
One option (lmk what you think): It might be cleaner not to try to mess with `transform_subclass` at all, and have the manual logic in meta_utils.py be something like:
```
if t.is_nested:
    attrs, ctx = t.__tensor_flatten__()
    transformed_tensors_dict = {}
    for attr in attrs:
        inner_t = getattr(t, attr)
        transformed_tensors_dict[attr] = callback(lambda: empty_create(inner_t, AttrSource(source, attr))
        assert isinstance(ctx, dict) and 'jagged_dim' in ctx
        # Replace the eager jagged dim with our freshly allocated jagged dim that has a source
        ctx['jagged_dim'] =  sym_ragged_size
        return type(t).__tensor_unflatten__(transformed_tensors_dict, ctx)
```
lambda attr, inner_t: callback(
                                lambda: empty_create(
                                    inner_t,
                                    AttrSource(source, attr),
                                )
                            ),

Reply:
Thanks for the suggestion, this seems like a cleaner option!

(From offline discussion, the ctx is a tuple today, but can be any object, e.g. dict since it is only used by the flatten/unflatten methods and treated as an opaque blob.)

Reply:
It has been updated now.